File: ref-impl/typescript/src/scope.ts
import { Scope, ScopeKind } from "./models.js";

/**
 * Normalize a file path:
 * - Remove leading ./
 * - Collapse multiple /
 * - Remove trailing /
 */
function normalizePath(path: string): string {
  // Remove leading ./
  let normalized = path.replace(/^\.\//, "");
  // Collapse multiple slashes
  normalized = normalized.replace(/\/+/g, "/");
  // Remove trailing slash (except for root)
  if (normalized !== "/" && normalized.endsWith("/")) {
    normalized = normalized.slice(0, -1);
  }
  return normalized;
}

/**
 * Check if two string sets have any intersection
 */
function setsIntersect(setA: string[], setB: string[]): boolean {
  const bSet = new Set(setB);
  for (const item of setA) {
    if (bSet.has(item)) {
      return true;
    }
  }
  return false;
}

/**
 * Check if two resources overlap (with path normalization)
 */
function resourcesOverlap(filesA: string[], filesB: string[]): boolean {
  const normalizedA = filesA.map(normalizePath);
  const normalizedB = filesB.map(normalizePath);
  return setsIntersect(normalizedA, normalizedB);
}

/**
 * Check if two entity sets overlap (exact match)
 */
function entitiesOverlap(entitiesA: string[], entitiesB: string[]): boolean {
  return setsIntersect(entitiesA, entitiesB);
}

/**
 * Check if two task id sets overlap (exact match)
 */
function taskIdsOverlap(tasksA: string[], tasksB: string[]): boolean {
  return setsIntersect(tasksA, tasksB);
}

/**
 * Detect if two scopes overlap
 * Returns true if there is any overlap, false if no overlap, conservative true for unknown kinds
 */
export function scopeOverlap(a: Scope, b: Scope): boolean {
  // Same kind comparison
  if (a.kind === b.kind) {
    switch (a.kind) {
      case ScopeKind.FILE_SET:
        if (a.resources && b.resources) {
          return resourcesOverlap(a.resources, b.resources);
        }
        break;
      case ScopeKind.ENTITY_SET:
        if (a.entities && b.entities) {
          return entitiesOverlap(a.entities, b.entities);
        }
        break;
      case ScopeKind.TASK_SET:
        if (a.task_ids && b.task_ids) {
          return taskIdsOverlap(a.task_ids, b.task_ids);
        }
        break;
    }
  }

  // Different kinds or unknown: conservative approach (assume overlap)
  return true;
}

/**
 * Check if *test* scope is fully contained within *container* scope.
 * Returns true when every item in test also appears in container.
 * For different scope kinds: conservative true (assume contained).
 */
export function scopeContains(container: Scope, test: Scope): boolean {
  if (container.kind !== test.kind) return true; // Conservative

  switch (container.kind) {
    case ScopeKind.FILE_SET: {
      const cSet = new Set((container.resources ?? []).map(normalizePath));
      return (test.resources ?? []).every((r) => cSet.has(normalizePath(r)));
    }
    case ScopeKind.ENTITY_SET: {
      const cSet = new Set(container.entities ?? []);
      return (test.entities ?? []).every((e) => cSet.has(e));
    }
    case ScopeKind.TASK_SET: {
      const cSet = new Set(container.task_ids ?? []);
      return (test.task_ids ?? []).every((t) => cSet.has(t));
    }
  }
  return true; // Unknown kind: conservative
}

/**
 * Check if scope is valid (has at least one non-empty set)
 */
export function isValidScope(scope: Scope): boolean {
  switch (scope.kind) {
    case ScopeKind.FILE_SET:
      return Array.isArray(scope.resources) && scope.resources.length > 0;
    case ScopeKind.ENTITY_SET:
      return Array.isArray(scope.entities) && scope.entities.length > 0;
    case ScopeKind.TASK_SET:
      return Array.isArray(scope.task_ids) && scope.task_ids.length > 0;
    default:
      return false;
  }
}

File: ref-impl/python/mpac/scope.py
"""Scope overlap detection for MPAC."""
from typing import List
import re

from .models import Scope


def normalize_path(path: str) -> str:
    """Normalize a file path for comparison.

    Steps:
    1. Remove leading ./
    2. Collapse multiple slashes (//)
    3. Remove trailing slashes

    Args:
        path: File path to normalize

    Returns:
        Normalized path
    """
    # Remove leading ./
    if path.startswith("./"):
        path = path[2:]

    # Collapse multiple slashes
    path = re.sub(r'/+', '/', path)

    # Remove trailing slash (except for root /)
    if path != "/" and path.endswith("/"):
        path = path.rstrip("/")

    return path


def scope_overlap(a: Scope, b: Scope) -> bool:
    """Detect if two scopes overlap.

    Args:
        a: First scope
        b: Second scope

    Returns:
        True if scopes overlap, False otherwise
    """
    # Different scope kinds: conservative assumption of overlap
    if a.kind != b.kind:
        return True

    # Same kind: check for intersection based on field names
    if a.kind == "file_set":
        # Normalize all paths and check for intersection
        a_items = {normalize_path(item) for item in (a.resources or [])}
        b_items = {normalize_path(item) for item in (b.resources or [])}
        return len(a_items & b_items) > 0

    elif a.kind == "entity_set":
        # Exact string matching for entity sets
        a_items = set(a.entities or [])
        b_items = set(b.entities or [])
        return len(a_items & b_items) > 0

    elif a.kind == "task_set":
        # Exact string matching for task sets
        a_items = set(a.task_ids or [])
        b_items = set(b.task_ids or [])
        return len(a_items & b_items) > 0

    else:
        # Unknown scope kind: conservative True
        return True


def scope_contains(container: Scope, test: Scope) -> bool:
    """Check if *test* scope is fully contained within *container* scope.

    Returns True when every item in *test* also appears in *container*.
    For different scope kinds: conservative True (assume contained).
    """
    if container.kind != test.kind:
        return True  # Conservative

    if container.kind == "file_set":
        c_items = {normalize_path(r) for r in (container.resources or [])}
        t_items = {normalize_path(r) for r in (test.resources or [])}
        return t_items.issubset(c_items)

    elif container.kind == "entity_set":
        return set(test.entities or []).issubset(set(container.entities or []))

    elif container.kind == "task_set":
        return set(test.task_ids or []).issubset(set(container.task_ids or []))

    return True  # Unknown kind: conservative
